cpp/include/cudf/io/text/data_chunk_source_factories.hpp

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/io/text/data_chunk_source.hpp>
+#include <cudf/scalar/scalar.hpp>
+
+#include <rmm/device_buffer.hpp>
+
+#include <thrust/host_vector.h>
+#include <thrust/system/cuda/experimental/pinned_allocator.h>
+
+#include <fstream>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+
+namespace cudf {
+namespace io {
+namespace text {
+
+namespace {
+
+/**
+ * @brief a reader which produces views of device memory which contain a copy of the data from an
+ * istream.
+ *
+ */
+class istream_data_chunk_reader : public data_chunk_reader {
+  struct host_ticket {
+    cudaEvent_t event;
+    thrust::host_vector<char, thrust::system::cuda::experimental::pinned_allocator<char>> buffer;
+  };
+
+ public:
+  istream_data_chunk_reader(std::unique_ptr<std::istream> datastream)
+    : _datastream(std::move(datastream)), _buffers(), _tickets(2)
+  {
+    // create an event to track the completion of the last device-to-host copy.
+    for (uint32_t i = 0; i < _tickets.size(); i++) {
+      CUDA_TRY(cudaEventCreate(&(_tickets[i].event)));
+    }
+  }
+
+  ~istream_data_chunk_reader()
+  {
+    for (uint32_t i = 0; i < _tickets.size(); i++) {
+      CUDA_TRY(cudaEventDestroy(_tickets[i].event));
+    }
+  }
+
+  device_span<char> find_or_create_data(uint32_t size, rmm::cuda_stream_view stream)
+  {
+    auto search = _buffers.find(stream.value());
```

Comment:
This can silently break in at least two scenarios. First, if `find_or_create_data` is called from two threads with their per-thread default stream. In both cases the `cudaStream_t`'s value will be 0, but they refer to different streams. 

Second, if a stream is created, used in `find_or_create_data`, destroyed, and a new stream is created and reuses the same stream ID, then once again you'll have a handle to the wrong stream.

Any time you're trying to use a stream ID as a key you run into these problems.

Reply:
I understand that this is an important thing to get right, especially if we decide to re-use it in other places in cuio. Right now the scope of this cache is limited to a single reading of the file, so the only invocations of this function will always see non-default streams from a stream pool, and once the work is done both the stream pool and cache will be destroyed.

How important is it to get this code reusable in this PR vs get it right the next time we try to use it? Would it be better to document the caveats and make a more general solution later?

Reply:
I don't see these being a problem in the current use case, and won't have a better solution until there are more use cases. Ok to resolve?

Reply:
I can get rid of the thread-id based lookup and use a concurrent collection to store unused tickets. if no tickets are available, one is created (up to a certain number). when a ticket is no longer in use, it can be returned to the collection. this way it doesn’t matter what stream is used to read the chunk.

Reply:
This is impossible to address differently unless we change `rmm`'s sync-and-steal behavior or use `cudaStreamAddCallback` to indicate when a buffer is ready for re-use.

Reply:
It can be resolved by mapping the user specified stream to an internal side stream and synchronizing the user specified stream with the internal stream. That way you know the lifetime of the internal streams.